src/main/scala/linting/rule/LintConflictingModuleNames.scala

```diff
@@ -0,0 +1,348 @@
+  def getName(desiredName: String)(module: Module): String = {
+    val noInfoModule = NamingStrategy.removeModuleInfo(module).copy(name = NamingStrategy.emptyName)
+    NamingStrategy.appendHashCode(desiredName, "C", noInfoModule.hashCode)
+  }
+}
+
+object NamingStrategy {
+
+  final val emptyName: String = ""
+
+  def appendHashCode(desiredName: String, hashPrefix: String, hashCode: Int): String = {
+    s"${desiredName}_$hashPrefix" + f"${hashCode}%08X"
+  }
+
+
+  // remove name helpers
+
+  def removeTypeNames(tpe: Type): Type = tpe match {
+    case g: GroundType => g
+    case v: VectorType => v
+    case b: BundleType => b.copy(fields = b.fields.map { field =>
+      field.copy(name = emptyName)
+    })
+  }
+
+  def removeStatementNames(stmt: Statement): Statement = stmt match {
+    case i: DefInstance => i.copy(
+      name = emptyName,
+      module = emptyName,
+      info = NoInfo)
+    case i: WDefInstance => i.copy(
+      name = emptyName,
+      module = emptyName,
+      tpe = removeTypeNames(i.tpe),
+      info = NoInfo)
+    case _ => stmt
+      .mapStmt(removeStatementNames)
+      .mapString(_ => emptyName)
+      .mapInfo(_ => NoInfo)
+      .mapType(removeTypeNames)
+  }
+
+  def removePortNames(port: Port): Port = {
+    port.copy(name = emptyName, tpe = removeTypeNames(port.tpe), info = NoInfo)
+  }
+
+  def removeModuleNames(mod: Module): Module = {
+    mod.copy(
+      ports = mod.ports.map(removePortNames(_)),
+      body = removeStatementNames(mod.body),
+      name = emptyName,
+      info = NoInfo)
+  }
+
+  // remove Info helpers
+
+  def removeStatementInfo(stmt: Statement): Statement = {
+    stmt.mapStmt(removeStatementInfo).mapInfo(_ => NoInfo)
+  }
+
+  def removePortInfo(port: Port): Port = {
+    port.copy(info = NoInfo)
+  }
+
+  def removeModuleInfo(mod: Module): Module = {
+    mod.copy(
+      ports = mod.ports.map(removePortInfo(_)),
+      body = removeStatementInfo(mod.body),
+      info = NoInfo)
+  }
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass,
+    PortStructureNamingStrategy.getClass,
+    ContentStructureNamingStrategy.getClass,
+    ContentNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies a desired name for a module
+  */
+case class ModuleNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): ModuleNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** An example of how to customize the names of certain modules using an Aspect
+  */
+case object StabilizeNamesAspect extends Aspect[RawModule] {
```

Comment:
Proposal: Remove `ModuleNameAnnotation`, replace with `DesiredNameAnnotation` and `OverrideDesiredNameAnnotation`.

We want the default to be to be able to detect name conflicts for all modules in our lint rule. To do this, we need all original desired names to be passed to the linter.

Additionally, we want to expose an Aspect API to let users override a module's desired name, in the case where a new name conflict arises and we don't want to have to change the module's source code (or the module lacks context). We want a transformation to consume these overrides to change the names in the FIRRTL circuit, and this transform deletes the original desired name and emits a new desired name annotation containing the override name.

Finally, we will have a linter which checks that all deduplications result in no name conflicts, where no module post deduplication shares its desired name annotation.